gateways/authorizenet/request_builders.go

```diff
@@ -109,3 +114,76 @@ func authentication(merchantName string, transactionKey string) MerchantAuthenti
 		TransactionKey: transactionKey,
 	}
 }
+
+func addL2L3Data(authRequest *sleet.AuthorizationRequest, authNetAuthRequest *CreateTransactionRequest) *CreateTransactionRequest {
+	if authRequest.Level3Data != nil {
+		authNetAuthRequest.TransactionRequest.LineItem = json.RawMessage(buildLineItemsString(authRequest))
+
+		authNetAuthRequest.TransactionRequest.Tax = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.TaxAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Duty = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.DutyAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Shipping = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.ShippingAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Customer = &Customer{
+			Id: authRequest.Level3Data.CustomerReference,
+		}
+	}
+
+	if authRequest.ShippingAddress != nil {
+		authNetAuthRequest.TransactionRequest.ShippingAddress = &BillingAddress{
+			FirstName: 	authRequest.CreditCard.FirstName,
+			LastName: 	authRequest.CreditCard.LastName,
+			Company: 	common.SafeStr(authRequest.ShippingAddress.Company),
+			Address: 	authRequest.ShippingAddress.StreetAddress1,
+			City:      	authRequest.ShippingAddress.Locality,
+			State:    	authRequest.ShippingAddress.RegionCode,
+			Zip:       	authRequest.ShippingAddress.PostalCode,
+			Country:   	authRequest.ShippingAddress.CountryCode,
+		}
+	}
+
+	return authNetAuthRequest
+}
+
+func buildLineItemsString(authRequest *sleet.AuthorizationRequest) string {
+	// No more than 30 lineitems
+	lineItemCount := len(authRequest.Level3Data.LineItems)
+	if lineItemCount > 30 {
+		lineItemCount = 30
+	}
+
+	lineItems := "{"
+	for i := 0; i < lineItemCount; i++ {
+		lineItem := &LineItem{
+			ItemId: sleet.TruncateString(authRequest.Level3Data.LineItems[i].CommodityCode, 31),
+			Name: sleet.TruncateString(authRequest.Level3Data.LineItems[i].ProductCode, 31),
+			Description: sleet.TruncateString(authRequest.Level3Data.LineItems[i].Description, 255),
+			Quantity: strconv.FormatInt(authRequest.Level3Data.LineItems[i].Quantity, 10),
+			UnitPrice: strconv.FormatInt(authRequest.Level3Data.LineItems[i].UnitPrice.Amount, 10),
+		}
+
+		// This is absolutely terrible why is json -> xml a thing
+		lineItemByte, err := json.Marshal(lineItem)
+		if err == nil {
+			// No error, add the string
+			// We must use a string because authnet allows invalid JSON
+			lineItems += "\"lineItem\":" + string(lineItemByte)
+			if i < lineItemCount - 1 {
+				lineItems += ","
+			}
+		}
+	}
+	lineItems += "}"
+
+	if len(lineItems) > 2 {
```

Comment:
Can we set a bool in the for loop rather than performing this check here? Why we are checking for len > 2 is a bit obscure at first glance.